### PR TITLE
fix(dates): ensure that server or client time is used when appropriate

### PR DESCRIPTION
**Pull request by Mohammer5**
### Changes
- [x] Refactors the `useDateStringAtServerTimezone` hook (see [`src/shared/date`](https://github.com/dhis2/aggregate-data-entry-app/tree/TECH-1328/src/shared/date))
- [x] Converts period start- and enddates to server time when they will be compared against start- and enddates of data sets or category options
- [x] Fixes tests that broke

### Questions:
- [x] I had to add the `useConvertClientDateAtServerTimezone` so a function can be provided to the metadata selectors. Is there a better alternative to doing it the way I've implemented it?
  - @amcgee & @HendrikThePendric proposed a better alternative
- [x] Should we re-use the `useConvertClientDateAtServerTimezone` in `useClientDateAtServerTimezone` to reduce duplication?
  - @amcgee & @HendrikThePendric proposed a better alternative

### CODE REVIEW

File: src/data-workspace/data-details-sidebar/basic-information.js [hunk 1]
```diff
@@ -2,68 +2,82 @@ import i18n from '@dhis2/d2-i18n'
...
-        <h1 className={styles.title}>
-            {item.displayFormName}
-            {item.categoryOptionComboName &&
-                ` | ${item.categoryOptionComboName}`}
-        </h1>
+const BasicInformation = ({ item }) => {
+    // This might pass "undefined" to moment and subsequently a wrong
+    // "timeAgo", but in that case we won't render anything anyway, so there's
+    // nothing to worry about in case there is no "item"
+    const lastUpdatedUtcDate = new Date(item.lastUpdated)
```
Comment (tomzemp) [inline src/data-workspace/data-details-sidebar/basic-information.js:14]: should the variable name be updated to not reference UTC?

**Review by tomzemp — CHANGES_REQUESTED** [review]
Looks good! Nice that the catOptions are also getting the server logic!

I like the approach of the closure returned by `useConvertClientDateAtServerTimezone`; to check: the selector should be cached regardless of this function, so we don't need a stable reference, right? I think it would be nice from a maintenance perspective if the more of the code was reused between `useConvertClientDateAtServerTimezone`, `useClientDateAtServerTimezone` and `useServerDateAtClientTimezone`, but that's not crucial.

I've left some individual comments, and also wondering if the audit-log created dates should be server or client time? (https://github.com/dhis2/aggregate-data-entry-app/blob/cecd87f1b6ce7b059f30dfa3d4d5a47762928b7f/src/data-workspace/data-details-sidebar/audit-log.js#L104)

**Comment by amcgee** [comment]
Sorry to butt in here, but I find these hooks names a bit hard to parse, which could lead to pretty subtle bugs if one is used incorrectly.

Here's how it would be used:

```jsx
function MyComponent() {
  const [clientServerDate, setClientServerDate] = useState(null)
  const { fromClientDate } = useClientServerDateUtils()
  
  useEffect(() => {
    const interval = setInterval(() => {
      setClientServerDate(fromClientDate(new Date())
    }, 1000)
    return () => { clearInterval(interval) }
  }, [])
  
  return <div>
    Client date: {clientServerDate.clientDate}<br>
    Server date: {clientServerDate.serverDate}
  </div>
}
```

And an implementation suggestion to make this a bit easier to parse when reading the code.  It might not be necessary to have a dedicated class, an object with two properties might also be OK, but it's nice to be standalone and strongly typed:

```js
class ClientServerDate {
    constructor(clientDate, serverDate) {
        this.clientDate = clientDate;
        this.serverDate = serverDate;
    }

    static fromServerDate(serverDate, serverTimeOffset) {
        return new ClientServerDate(new Date(serverDate - serverTimeOffset), serverDate)
    }

    static fromClientDate(clientDate, serverTimeOffset) {
        return new ClientServerDate(clientDate, new Date(clientDate + serverTimeOffset))
    }
}

const useServerTimeOffset = () => {
    const { systemInfo } = useConfig()

    return useMemo(() => {
        const currentDate = getCurrentDate()
        const nowAtServerTimeZone = new Date(
            currentDate.toLocaleString('en-US', {
                timeZone: systemInfo.serverTimeZoneId,
            })
        )

        const currentDateTime = currentDate.getTime()
        const nowAtServerTimeZoneTime = nowAtServerTimeZone.getTime()
        return nowAtServerTimeZoneTime - currentDateTime
    }, [systemInfo])
}

const useClientServerDateUtils = () => {
    const serverTimeOffset = useServerTimeOffset()
    return {
        fromServerDate: (serverTime) => ClientServerDate.fromServerDate(serverTime, serverTimeOffset),
        fromClientDate: (clientTime) => ClientServerDate.fromClientDate(clientTime, serverTimeOffset)
    }
}
```

**Comment by HendrikThePendric** [comment]
@amcgee sorry for ignoring your suggestion completely. I see we both posted around the same time, I think that is why I didn't see yours. I think having a stable hook which returns conversion utility functions probably makes a lot of sense. 

The only thing against it is that if we keep recomputing the date on every render then we it'd work well for when the user happens to pass a date boundary while working in an app: if this string was a dependency of a `useEffect` hook, the effect would run on the first render after crossing the date boundary, because a different date string would be returned. But I think it's fairly safe to say this is pretty much literally an edge case 🤣 . Overall the API you came up with is more readable and versatile.

**Comment by amcgee** [comment]
> The only thing against it is that if we keep recomputing the date on every render then we it'd work well for when the user happens to pass a date boundary while working in an app: if this string was a dependency of a useEffect hook, the effect would run on the first render after crossing the date boundary, because a different date string would be returned. But I think it's fairly safe to say this is pretty much literally an edge case 🤣 . Overall the API you came up with is more readable and versatile.

Hmm yeah that is an interesting edge-case... I think requiring a page refresh when crossing into a new timezone is probably a fair complexity trade-off 😆

**Comment by amcgee** [comment]
> @amcgee sorry for ignoring your suggestion completely. I see we both posted around the same time, I think that is why I didn't see yours. I think having a stable hook which returns conversion utility functions probably makes a lot of sense.

No worries!  I see we both recommended the same `useServerTimeOffset` type hook :-)  That seems like a nice refactor to start things off.

I like a simple, easy-to-understand API for this - the one I suggested is decent but could probably be improved further I think...

File: src/data-workspace/data-details-sidebar/basic-information.js (hunk 1 above)
Comment (Mohammer5) [reply src/data-workspace/data-details-sidebar/basic-information.js]: addressed

File: src/context-selection/attribute-option-combo-selector-bar-item/use-category-combination.js [hunk 2]
```diff
@@ -32,6 +37,8 @@ const isOptionWithinPeriod = ({
 }
 
 export default function useCategoryCombination() {
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/context-selection/attribute-option-combo-selector-bar-item/use-category-combination.js]: Removed

File: src/context-selection/period-selector-bar-item/use-date-limit.js [hunk 3]
```diff
@@ -45,15 +48,17 @@ export const computeDateLimit = ({
...
+        getCurrentDate()
+    )
 
     return useMemo(
         () =>
             computeDateLimit({
                 dataSetId,
                 metadata,
-                adjustedCurrentDateString,
+                currentDateAtServerTimezone,
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/context-selection/period-selector-bar-item/use-date-limit.js]: Addressed, see comment in the file above the hook's dependency array

**Comment by Mohammer5** [comment]
@amcgee @HendrikThePendric I've change the implementation and followed @amcgee's example closely (and marked you as a co-author in the commit). I've also added tests and updated the code base to use those utils

**Review by tomzemp — APPROVED** [review]
Overall looks good and I like the new approach. I think it could be useful (but defer to @Mohammer5):

- to get rid of `setClientDate` and `setServerDate` in use-client-server-date as we're not using this code, and I don't think we have a use case in this app for updating the date/time?
- remove the simplification we had in use-check-lock-status (where we were checking based on the date rather than the date-time), as we now have fromClientDate we could use?
- consider not naming the objects returned from `useClientServerDate` as `clientServerDate`. For me, this becomes a bit confusing, like in basic-information where we declare `const clientServerDate = useClientServerDate({ serverDate: lastUpdated })` and then have `const timeAgo = moment(clientServerDate.clientDate).fromNow()`; I think it might be more understandable if we gave the object a meaningful name like `const lastUpdatedDate = useClientServerDate({ serverDate: lastUpdated })` so that when you then make a comparison `moment(lastUpdatedDate.clientDate).fromNow()` it's clearer that you're taking the lastUpdated date in client time?

**Comment by Mohammer5** [comment]
> * to get rid of setClientDate and setServerDate in use-client-server-date as we're not using this code, and I don't think we have a use case in this app for updating the date/time?

:ok: 

> * remove the simplification we had in use-check-lock-status (where we were checking based on the date rather than the date-time), as we now have fromClientDate we could use?

Not sure what you mean. We're passing the current day string as dependency instead of the date so the hook doesn't recompute every time it rerenders (as the current date will always change). I've added a comment somewhere explaining that. I don't think that this is a show stopper and as you've approved already, I'll merge this PR. We can improve things later anyway. If you want to talk about it, feel free & hit me up on Slack!

> * consider not naming the objects returned from useClientServerDate as clientServerDate. For me, this becomes a bit confusing, like in basic-information where we declare const clientServerDate = useClientServerDate({ serverDate: lastUpdated }) and then have const timeAgo = moment(clientServerDate.clientDate).fromNow(); I think it might be more understandable if we gave the object a meaningful name like const lastUpdatedDate = useClientServerDate({ serverDate: lastUpdated }) so that when you then make a comparison moment(lastUpdatedDate.clientDate).fromNow() it's clearer that you're taking the lastUpdated date in client time?

:ok: I've chosen proper names for everything and made it a lot simpler
